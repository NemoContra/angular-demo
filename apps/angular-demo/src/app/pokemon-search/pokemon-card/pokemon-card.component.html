@let cardPokemon = pokemon();
<!-- Give the card a border with color #e3350d if it is selected -->
<div class="card">
  <div class="content">
    <!-- Render the name of the pokemon in the h2 and format it in titlecase -->
    <h2>Name here in Titlecase</h2>
    <img
      [ngSrc]="cardPokemon.sprites.front_default"
      [alt]="cardPokemon.name"
      height="96"
      width="96"
    />

    <dl class="details">
      <dt>Type</dt>
      <dd>
        <!-- List all the types of the pokemon (type.name), separate them by comma -->
        Fill in types here
      </dd>
      <dt>Pokédex No.</dt>
      <dd>{{ cardPokemon.id }}</dd>
      <dt>Height</dt>
      <dd>{{ cardPokemon.height / 10 | number : '1.0-2' }} m</dd>
      <dt>Weight</dt>
      <dd>{{ cardPokemon.weight / 10 | number : '1.0-2' }} kg</dd>
    </dl>
  </div>
  <div class="actions">
    <!-- Toggle the selection of the pokemon when the button is clicked -->
    <button type="button">
      <!-- Render 'Remove Pokémon' or 'Add Pokémon' base on the state of selection  -->
      Button
    </button>
  </div>
</div>
