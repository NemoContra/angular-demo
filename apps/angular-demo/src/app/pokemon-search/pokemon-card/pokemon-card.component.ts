import {
  ChangeDetectionStrategy,
  Component,
  input,
  model,
} from '@angular/core';
import { Pokemon } from '../../models/pokemon';
import { DecimalPipe, NgOptimizedImage } from '@angular/common';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [NgOptimizedImage, DecimalPipe],
  selector: 'pokemon-card',
  styleUrl: './pokemon-card.component.scss',
  templateUrl: './pokemon-card.component.html',
})
export class PokemonCardComponent {
  pokemon = input.required<Pokemon>();
  selected = model<boolean>(false);

  // PBI #2: Select and deselect
  toggleSelected() {
    this.selected.update((selected) => !selected);
  }
}
