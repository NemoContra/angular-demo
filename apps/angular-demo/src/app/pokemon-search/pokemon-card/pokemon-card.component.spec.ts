import { ComponentFixture, TestBed } from '@angular/core/testing';
import { PokemonCardComponent } from './pokemon-card.component';
import { mockPokemons } from '../../testing/mock-data/pokemons';

describe('PokemonCardComponent', () => {
  let component: PokemonCardComponent;
  let fixture: ComponentFixture<PokemonCardComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [PokemonCardComponent],
    }).compileComponents();

    fixture = TestBed.createComponent(PokemonCardComponent);
    component = fixture.componentInstance;
    fixture.componentRef.setInput('pokemon', mockPokemons[0]);
    fixture.componentRef.setInput('selected', false);
    fixture.detectChanges();
  });

  it.todo('should render with initial inputs');

  it('should select the pokemon', () => {
    fixture.componentRef.setInput('selected', true);
    // detect changes
    // query the button text
    const buttonText = '';
    expect(buttonText).toEqual('Remove Pokémon');
  });

  it('should deselect the pokemon', () => {
    fixture.componentRef.setInput('selected', false);
    // detect changes
    // query the button text
    const buttonText = '';
    expect(buttonText).toEqual('Add Pokémon');
  });

  it('should toggle the selected state', () => {
    const selectedSpy = null as any; // create a spy
    component.selected.subscribe(selectedSpy);

    // expect that spy has not been called yet
    // interaction
    // expect that spy has been called with boolean value true
  });
});
