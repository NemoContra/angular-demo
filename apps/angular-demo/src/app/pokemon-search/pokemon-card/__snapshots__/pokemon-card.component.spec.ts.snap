// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCardComponent should render with initial inputs 1`] = `
<pokemon-card
  pokemon={[Function Function]}
  selected={[Function Function]}
>
  <div
    class="card"
  >
    <div
      class="content"
    >
      <h2>
        Bulbasaur
      </h2>
      <img
        alt="bulbasaur"
        fetchpriority="auto"
        height="96"
        loading="lazy"
        ng-img="true"
        src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
        width="96"
      />
      <dl
        class="details"
      >
        <dt>
          Type
        </dt>
        <dd>
           Grass
          , 
           Poison
        </dd>
        <dt>
          Pokédex No.
        </dt>
        <dd>
          1
        </dd>
        <dt>
          Height
        </dt>
        <dd>
          0.6 m
        </dd>
        <dt>
          Weight
        </dt>
        <dd>
          6.9 kg
        </dd>
      </dl>
    </div>
    <div
      class="actions"
    >
      <button
        type="button"
      >
         Add Pokémon 
      </button>
    </div>
  </div>
</pokemon-card>
`;
